tm2/pkg/sdk/bank/keeper.go

```diff
@@ -84,8 +148,42 @@ func (bank BankKeeper) InputOutputCoins(ctx sdk.Context, inputs []Input, outputs
 	return nil
 }
 
-// SendCoins moves coins from one account to another
+// canSendCoins returns true if the coins can be sent without violating any restriction.
+func (bank BankKeeper) canSendCoins(ctx sdk.Context, addr crypto.Address, amt std.Coins) bool {
+	if len(bank.restrictedDenoms) == 0 {
+		// No restrictions.
+		return true
+	}
+	if amt.ContainOneOfDenom(bank.restrictedDenoms) {
+		if acc := bank.acck.GetAccount(ctx, addr); acc != nil && acc.IsRestricted() {
+			return false
+		}
+	}
+	return true
+}
+
+// SendCoins moves coins from one account to another, restrction could be applied
 func (bank BankKeeper) SendCoins(ctx sdk.Context, fromAddr crypto.Address, toAddr crypto.Address, amt std.Coins) error {
+	return bank.sendCoins(ctx, fromAddr, toAddr, amt, true)
+}
+
+func (bank BankKeeper) SendCoinsUnrestricted(ctx sdk.Context, fromAddr crypto.Address, toAddr crypto.Address, amt std.Coins) error {
+	return bank.sendCoins(ctx, fromAddr, toAddr, amt, false)
+}
+
+func (bank BankKeeper) sendCoins(
+	ctx sdk.Context,
+	fromAddr crypto.Address,
+	toAddr crypto.Address,
+	amt std.Coins,
+	wRestriction bool, // with restriction
```

Comment:
Nitpick: this `wRestriction` bool should not be propagated to `sendCoins`, but instead, `sendCoins` can simply do the subtract + add, and you handle the restriction logic in `SendCoins`:

```go
func (bank BankKeeper) SendCoins(ctx sdk.Context, fromAddr crypto.Address, toAddr crypto.Address, amt std.Coins) error {
	if !bank.canSendCoins(ctx, fromAddr, amt) {
		return std.RestrictedTransferError{}
	}

	return bank.sendCoins(ctx, fromAddr, toAddr, amt)
}
```